File: cpp/tests/test_cluster.cpp
// S4 on two blobs far enough apart that no chain of 6-connected voxels joins
// them. The spec asks for the component with the most *points*, not the most
// voxels, so one fixture makes the two differ.

#include <cstdint>
#include <vector>

#include "grasp_core/cluster.hpp"
#include "grasp_core/config.hpp"

#include "check.hpp"

using grasp_core::Clusterer;
using grasp_core::Config;

namespace
{

void add_blob(
  std::vector<float> & points, double cx, double cy, double cz, double half, int count)
{
  for (int i = 0; i < count; ++i) {
    // A deterministic lattice rather than noise: the fixture must not depend
    // on a generator's stream position.
    const double t = static_cast<double>(i) / static_cast<double>(count);
    points.push_back(static_cast<float>(cx + half * (2.0 * t - 1.0)));
    points.push_back(static_cast<float>(cy + half * (2.0 * ((3 * i) % 7) / 6.0 - 1.0)));
    points.push_back(static_cast<float>(cz + half * (2.0 * ((5 * i) % 11) / 10.0 - 1.0)));
  }
}

void test_picks_the_blob_with_more_points(const Config & config)
{
  std::vector<float> points;
  const int small_count = 300;
  const int large_count = 900;
  add_blob(points, 0.30, -0.15, 0.05, 0.015, small_count);
  add_blob(points, 0.55, 0.18, 0.05, 0.015, large_count);
  const int n = small_count + large_count;

  Clusterer clusterer(config);
  clusterer.reserve(static_cast<std::size_t>(n));
  std::vector<int> indices(static_cast<std::size_t>(n));
  const int count = clusterer.run(points.data(), n, indices.data());

  CHECK(count == large_count);
  // The winning component is the second blob, so every index is past the first.
  for (int i = 0; i < count; ++i) {
    CHECK(indices[static_cast<std::size_t>(i)] >= small_count);
  }
  // Indices come back ascending, which is what makes the downstream reductions
  // order-independent between the two implementations.
  for (int i = 1; i < count; ++i) {
    CHECK(indices[static_cast<std::size_t>(i - 1)] < indices[static_cast<std::size_t>(i)]);
  }
}

void test_more_voxels_does_not_beat_more_points(const Config & config)
{
  // A sparse sheet spanning many voxels against a dense cube in few. Point
  // count decides, so the cube wins.
  std::vector<float> points;
  const double voxel = config.cluster.voxel_size_m;
  int sheet_count = 0;
  for (int i = 0; i < 60; ++i) {
    points.push_back(static_cast<float>(0.10 + static_cast<double>(i) * voxel));
    points.push_back(0.30f);
    points.push_back(0.10f);
    ++sheet_count;
  }
  const int cube_count = 400;
  add_blob(points, 0.50, -0.20, 0.08, 0.008, cube_count);

  Clusterer clusterer(config);
  const int n = sheet_count + cube_count;
  clusterer.reserve(static_cast<std::size_t>(n));
  std::vector<int> indices(static_cast<std::size_t>(n));
  const int count = clusterer.run(points.data(), n, indices.data());

  CHECK(count == cube_count);
  CHECK(indices[0] == sheet_count);
}

void test_below_min_points_returns_empty(const Config & config)
{
  std::vector<float> points;
  const int count = config.cluster.min_points - 1;
  add_blob(points, 0.40, 0.0, 0.05, 0.005, count);

  Clusterer clusterer(config);
  clusterer.reserve(static_cast<std::size_t>(count));
  std::vector<int> indices(static_cast<std::size_t>(count));
  CHECK(clusterer.run(points.data(), count, indices.data()) == 0);
}

void test_six_connectivity_does_not_bridge_a_diagonal(const Config & config)
{
  // Two single-voxel groups touching only at a corner. Under 26-connectivity
  // they would merge; under the 6-connectivity the spec fixes they must not,
  // and with both below min_points the answer is empty.
  const double v = config.cluster.voxel_size_m;
  std::vector<float> points;
  const int per_voxel = config.cluster.min_points - 1;
  for (int i = 0; i < per_voxel; ++i) {
    points.push_back(static_cast<float>(0.5 * v));
    points.push_back(static_cast<float>(0.5 * v));
    points.push_back(static_cast<float>(0.5 * v));
  }
  for (int i = 0; i < per_voxel; ++i) {
    points.push_back(static_cast<float>(1.5 * v));
    points.push_back(static_cast<float>(1.5 * v));
    points.push_back(static_cast<float>(1.5 * v));
  }

  Clusterer clusterer(config);
  const int n = 2 * per_voxel;
  clusterer.reserve(static_cast<std::size_t>(n));
  std::vector<int> indices(static_cast<std::size_t>(n));
  CHECK(clusterer.run(points.data(), n, indices.data()) == 0);
}

}  // namespace

int main()
{
  const Config config = Config::load(grasp_test::asset("assets/pipeline_config.json"));
  test_picks_the_blob_with_more_points(config);
  test_more_voxels_does_not_beat_more_points(config);
  test_below_min_points_returns_empty(config);
  test_six_connectivity_does_not_bridge_a_diagonal(config);
  return grasp_test::report("test_cluster");
}

File: python/tests/test_cluster.py
"""Connected components on blobs whose membership is known by construction.

The interesting cases are the ones where the answer is decided by a rule rather
than by the data: most points rather than most voxels, and the lexicographic
tie-break. Both are places where a hash-map implementation and a dense-grid one
can silently disagree.
"""
import numpy as np
import pytest


@pytest.fixture(scope='module')
def clusterer(config):
    from grasp_core.cluster import Clusterer
    voxel = config['cluster']['voxel_size_m']
    workspace = config['workspace']
    span = 1
    for axis in ('x', 'y', 'z'):
        span *= int(np.ceil((workspace[f'{axis}_max']
                             - workspace[f'{axis}_min']) / voxel)) + 2
    clusterer = Clusterer(config['cluster'], span)
    clusterer.resize(20000)
    return clusterer


def blob(centre, count, radius, seed):
    rng = np.random.default_rng(seed)
    return np.asarray(centre) + rng.uniform(-radius, radius, size=(count, 3))


def test_picks_the_blob_with_the_most_points(clusterer, config):
    small = blob((0.30, -0.10, 0.05), config['cluster']['min_points'] * 2,
                 0.015, 1)
    large = blob((0.55, 0.12, 0.05), config['cluster']['min_points'] * 5,
                 0.015, 2)
    points = np.vstack([small, large])
    indices = clusterer.run(points)
    assert indices.size == large.shape[0]
    assert np.array_equal(np.sort(indices),
                          np.arange(small.shape[0], points.shape[0]))


def test_most_points_beats_most_voxels(clusterer, config):
    """A dense clump of few voxels must beat a sparse shell of many."""
    voxel = config['cluster']['voxel_size_m']
    minimum = config['cluster']['min_points']
    rng = np.random.default_rng(3)
    dense = np.array([0.30, 0.0, 0.05]) + rng.uniform(
        0.0, voxel * 0.9, size=(minimum * 4, 3))
    # A connected line of voxels, one point each: many voxels, few points.
    sparse = np.stack([
        np.array([0.55, 0.0, 0.05]) + np.array([i * voxel, 0.0, 0.0])
        for i in range(minimum * 2)])
    points = np.vstack([dense, sparse])
    indices = clusterer.run(points)
    assert indices.size == dense.shape[0]
    assert indices.max() < dense.shape[0]


def test_tie_breaks_toward_the_lowest_key(clusterer, config):
    """Two components with identical point counts: the one holding the
    lexicographically smallest (kx, ky, kz) wins."""
    voxel = config['cluster']['voxel_size_m']
    minimum = config['cluster']['min_points']
    rng = np.random.default_rng(5)
    offsets = rng.uniform(0.0, voxel * 0.5, size=(minimum * 2, 3))
    low = np.array([0.20, -0.20, 0.05]) + offsets
    high = np.array([0.60, 0.20, 0.05]) + offsets
    # Whichever order the points arrive in, the same component must win.
    for order in (np.vstack([low, high]), np.vstack([high, low])):
        indices = clusterer.run(order)
        assert indices.size == minimum * 2
        centre = order[indices].mean(axis=0)
        assert centre[0] < 0.4 and centre[1] < 0.0


def test_returns_empty_below_the_minimum(clusterer, config):
    points = blob((0.4, 0.0, 0.05), config['cluster']['min_points'] - 1,
                  0.01, 7)
    assert clusterer.run(points).size == 0


def test_six_connectivity_does_not_join_diagonal_voxels(clusterer, config):
    """Face neighbours only: two clumps touching at a corner stay apart."""
    voxel = config['cluster']['voxel_size_m']
    minimum = config['cluster']['min_points']
    rng = np.random.default_rng(9)
    base = np.array([0.40, 0.0, 0.05])
    first = base + rng.uniform(0.0, voxel * 0.9, size=(minimum * 3, 3))
    diagonal = base + np.array([voxel, voxel, 0.0]) \
        + rng.uniform(0.0, voxel * 0.9, size=(minimum * 2, 3))
    indices = clusterer.run(np.vstack([first, diagonal]))
    assert indices.size == first.shape[0]
